File: typescript/src/dependencyAnalyzer.ts
/**
 * DependencyAnalyzer — detects data dependencies between tool calls.
 *
 * A tool call B depends on A when:
 *  1. B.dependsOn includes A's id (explicit declaration), OR
 *  2. Any string value inside B.inputs matches "$results.<A.id>[.<path>]"
 *     (implicit, reference-based dependency).
 */

/** Pattern matching $results.<toolId>[.<path>] placeholders. */
const REF_PATTERN = /\$results\.([a-zA-Z0-9_\-.\[\]]+)/g;

/** A single planned tool call. */
export interface ToolCall {
  /** Unique identifier within the execution plan. */
  id: string;
  /** Tool/function name to invoke. */
  name: string;
  /** Arguments to pass to the tool. */
  inputs?: Record<string, unknown>;
  /** Explicit list of call ids this call must wait for. */
  dependsOn?: string[];
  /** Per-call timeout in milliseconds (overrides orchestrator default). */
  timeoutMs?: number;
  /** Number of retry attempts on transient failure. */
  retries?: number;
  /** Arbitrary metadata — not used for dependency detection. */
  metadata?: Record<string, unknown>;
}

/** Result of the dependency analysis phase. */
export interface DependencyGraph {
  /** All tool calls, keyed by id. */
  nodes: Map<string, ToolCall>;
  /** id → set of ids it depends on. */
  edges: Map<string, Set<string>>;
  /** id → set of ids that depend on it (reverse). */
  revEdges: Map<string, Set<string>>;
}

function resolveRefToToolId(rawRef: string, knownIds: Set<string>): string {
  if (knownIds.has(rawRef)) return rawRef;

  let candidate = rawRef;
  while (candidate.includes(".")) {
    candidate = candidate.substring(0, candidate.lastIndexOf("."));
    if (knownIds.has(candidate)) return candidate;
  }

  return rawRef;
}

/** Returns all $results.<id> references found anywhere inside a value. */
function extractRefs(value: unknown, knownIds: Set<string>): Set<string> {
  const refs = new Set<string>();
  if (typeof value === "string") {
    let m: RegExpExecArray | null;
    // Reset lastIndex before each iteration
    REF_PATTERN.lastIndex = 0;
    while ((m = REF_PATTERN.exec(value)) !== null) {
      refs.add(resolveRefToToolId(m[1], knownIds));
    }
  } else if (Array.isArray(value)) {
    for (const item of value) {
      for (const ref of extractRefs(item, knownIds)) refs.add(ref);
    }
  } else if (value !== null && typeof value === "object") {
    for (const v of Object.values(value as Record<string, unknown>)) {
      for (const ref of extractRefs(v, knownIds)) refs.add(ref);
    }
  }
  return refs;
}

/** Detect cycles in the dependency graph using DFS. */
function isAcyclic(edges: Map<string, Set<string>>): boolean {
  const visited = new Set<string>();
  const stack = new Set<string>();

  function dfs(node: string): boolean {
    visited.add(node);
    stack.add(node);
    for (const dep of edges.get(node) ?? []) {
      if (!visited.has(dep)) {
        if (!dfs(dep)) return false;
      } else if (stack.has(dep)) {
        return false;
      }
    }
    stack.delete(node);
    return true;
  }

  for (const node of edges.keys()) {
    if (!visited.has(node)) {
      if (!dfs(node)) return false;
    }
  }
  return true;
}

/** Analyzes a list of ToolCalls and builds a DependencyGraph. */
export class DependencyAnalyzer {
  analyze(calls: ToolCall[]): DependencyGraph {
    const knownIds = new Set(calls.map((c) => c.id));
    const nodes = new Map(calls.map((c) => [c.id, c]));
    const edges = new Map<string, Set<string>>(calls.map((c) => [c.id, new Set()]));
    const revEdges = new Map<string, Set<string>>(calls.map((c) => [c.id, new Set()]));

    for (const call of calls) {
      const deps = new Set<string>(call.dependsOn ?? []);

      // Add implicit deps from $results.* references
      for (const ref of extractRefs(call.inputs ?? {}, knownIds)) {
        deps.add(ref);
      }

      // Validate all referenced ids exist
      for (const dep of deps) {
        if (!knownIds.has(dep)) {
          throw new Error(
            `ToolCall '${call.id}' references unknown tool id '${dep}'`
          );
        }
      }

      // Self-dependency is a no-op
      deps.delete(call.id);

      edges.set(call.id, deps);
      for (const dep of deps) {
        revEdges.get(dep)!.add(call.id);
      }
    }

    const graph: DependencyGraph = { nodes, edges, revEdges };

    if (!isAcyclic(edges)) {
      throw new Error(
        "Circular dependency detected in tool call plan. Synapse requires an acyclic graph."
      );
    }

    return graph;
  }
}

File: python/synapse/dependency_analyzer.py
"""
Dependency Analyzer — detects data dependencies between tool calls.

A tool call B depends on A if any input value of B references an output
placeholder produced by A (e.g. "$results.tool_a.output_field").
We also support explicit `depends_on` declarations for cases where the
dependency is semantic rather than syntactic.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any

# Pattern that matches output references like $results.<tool_id>.<field>
_REF_PATTERN = re.compile(r"\$results\.([a-zA-Z0-9_\-]+)")


@dataclass
class ToolCall:
    """
    Represents a single planned tool call.

    Attributes:
        id:         Unique identifier for this call within an execution plan.
        name:       Tool/function name to invoke.
        inputs:     Keyword arguments passed to the tool.
        depends_on: Explicit list of tool-call IDs this call must wait for.
                    Synapse also detects implicit deps from $results.* refs.
        timeout:    Per-call timeout in seconds (None = global default).
        retries:    Number of retry attempts on transient failure.
        metadata:   Arbitrary caller-supplied metadata (not used for deps).
    """

    id: str
    name: str
    inputs: dict[str, Any] = field(default_factory=dict)
    depends_on: list[str] = field(default_factory=list)
    timeout: float | None = None
    retries: int = 0
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass
class DependencyGraph:
    """
    Result of dependency analysis.

    Attributes:
        nodes:      All ToolCall objects, keyed by id.
        edges:      Mapping from each call id → set of ids it depends on.
        rev_edges:  Reverse mapping: id → set of ids that depend on it.
    """

    nodes: dict[str, ToolCall]
    edges: dict[str, set[str]]
    rev_edges: dict[str, set[str]]

    def dependencies_of(self, call_id: str) -> set[str]:
        return self.edges.get(call_id, set())

    def dependents_of(self, call_id: str) -> set[str]:
        return self.rev_edges.get(call_id, set())

    def roots(self) -> list[str]:
        """Tool calls with no dependencies — can start immediately."""
        return [cid for cid, deps in self.edges.items() if not deps]

    def is_acyclic(self) -> bool:
        """Return True if the graph contains no cycles (valid DAG)."""
        visited: set[str] = set()
        stack: set[str] = set()

        def dfs(node: str) -> bool:
            visited.add(node)
            stack.add(node)
            for dep in self.edges.get(node, set()):
                if dep not in visited:
                    if not dfs(dep):
                        return False
                elif dep in stack:
                    return False
            stack.discard(node)
            return True

        return all(dfs(n) for n in self.nodes if n not in visited)


def _extract_refs(value: Any) -> set[str]:
    """
    Recursively walk a value (dict, list, str, …) and collect all
    tool-call IDs referenced via $results.<id> syntax.
    """
    refs: set[str] = set()
    if isinstance(value, str):
        for match in _REF_PATTERN.finditer(value):
            refs.add(match.group(1))
    elif isinstance(value, dict):
        for v in value.values():
            refs |= _extract_refs(v)
    elif isinstance(value, (list, tuple)):
        for item in value:
            refs |= _extract_refs(item)
    return refs


class DependencyAnalyzer:
    """
    Analyzes a list of ToolCall objects and builds a DependencyGraph.

    Dependency detection rules (applied in order, union of all matches):
    1. Explicit ``depends_on`` field on the ToolCall.
    2. ``$results.<tool_id>`` placeholder references inside ``inputs``.

    Raises:
        ValueError: If a referenced tool id does not exist in the plan,
                    or if the resulting graph contains a cycle.
    """

    def analyze(self, calls: list[ToolCall]) -> DependencyGraph:
        known_ids = {c.id for c in calls}
        nodes = {c.id: c for c in calls}
        edges: dict[str, set[str]] = {c.id: set() for c in calls}
        rev_edges: dict[str, set[str]] = {c.id: set() for c in calls}

        for call in calls:
            deps: set[str] = set(call.depends_on)
            deps |= _extract_refs(call.inputs)

            # Validate referenced IDs exist
            unknown = deps - known_ids
            if unknown:
                raise ValueError(
                    f"ToolCall '{call.id}' references unknown tool ids: {unknown}"
                )

            # A call cannot depend on itself
            deps.discard(call.id)

            edges[call.id] = deps
            for dep_id in deps:
                rev_edges[dep_id].add(call.id)

        graph = DependencyGraph(nodes=nodes, edges=edges, rev_edges=rev_edges)

        if not graph.is_acyclic():
            raise ValueError(
                "Circular dependency detected in tool call plan. "
                "Synapse requires an acyclic dependency graph."
            )

        return graph
